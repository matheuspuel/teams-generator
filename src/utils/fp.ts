export * as Apply from 'fp-ts/Apply'
export * as A from 'fp-ts/Array'
export * as Bool from 'fp-ts/boolean'
export * as Console from 'fp-ts/Console'
export * as E from 'fp-ts/Either'
export * as Eq from 'fp-ts/Eq'
export {
  absurd,
  apply,
  constant,
  constFalse,
  constNull,
  constTrue,
  constUndefined,
  constVoid,
  flow,
  identity,
  Lazy,
  pipe,
  pipe as $,
  flow as $f,
} from 'fp-ts/function'
export * as IOE from 'fp-ts/IOEither'
export * as IOO from 'fp-ts/IOOption'
export * as Json from 'fp-ts/Json'
export * as Monoid from 'fp-ts/Monoid'
export * as NEA from 'fp-ts/NonEmptyArray'
export * as Num from 'fp-ts/number'
export * as O from 'fp-ts/Option'
export * as R from 'fp-ts/Reader'
export * as RT from 'fp-ts/ReaderTask'
export * as RTE from 'fp-ts/ReaderTaskEither'
export * as RA from 'fp-ts/ReadonlyArray'
export * as RNEA from 'fp-ts/ReadonlyNonEmptyArray'
export * as RR from 'fp-ts/ReadonlyRecord'
export * as Rec from 'fp-ts/Record'
export * as Semigroup from 'fp-ts/Semigroup'
export * as S from 'fp-ts/State'
export * as Show from 'fp-ts/Show'
export * as SRTE from 'fp-ts/StateReaderTaskEither'
export * as Str from 'fp-ts/string'
export * as T from 'fp-ts/Task'
export * as TE from 'fp-ts/TaskEither'
export * as TO from 'fp-ts/TaskOption'
export * as TT from 'fp-ts/TaskThese'
export * as Th from 'fp-ts/These'
export * as Tup from 'fp-ts/Tuple'
export * as D from 'io-ts/Decoder'
import { Either, Left, left, Right, right } from 'fp-ts/Either'
import * as IO_ from 'fp-ts/IO'
import { None, none, Option, Some, some } from 'fp-ts/Option'
import * as Ord_ from 'fp-ts/Ord'
import { Reader } from 'fp-ts/Reader'
import { ReaderTaskEither } from 'fp-ts/ReaderTaskEither'
import { State } from 'fp-ts/State'
import { StateReaderTaskEither } from 'fp-ts/StateReaderTaskEither'
import { Both, both, These } from 'fp-ts/These'
import { Decoder } from 'io-ts/Decoder'

export { Either, Left, left, Right, right }
export { None, none, Option, Some, some }
export { Reader }
export { ReaderTaskEither }
export { State }
export { StateReaderTaskEither }
export { Both, both, These }
export { Decoder }

export const IO = IO_
export type IO<A> = IO_.IO<A>
export const Ord = Ord_
export type Ord<A> = Ord_.Ord<A>
