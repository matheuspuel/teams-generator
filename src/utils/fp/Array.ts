import { constant, pipe } from 'effect/Function'
import * as O from 'effect/Option'
import { Option } from 'effect/Option'
import * as A from 'effect/ReadonlyArray'
import { NonEmptyReadonlyArray } from 'effect/ReadonlyArray'
import * as Eq from './Equivalence'

export * from 'effect/ReadonlyArray'

export const isArray = <A>(a: A): a is Extract<A, ReadonlyArray<unknown>> =>
  Array.isArray(a)

export const toNonEmpty = <A>(
  array: ReadonlyArray<A>,
): Option<NonEmptyReadonlyArray<A>> =>
  A.match(array, { onEmpty: () => O.none(), onNonEmpty: O.some })

export const getUnorderedEquivalence = <A>(
  E: Eq.Equivalence<A>,
): Eq.Equivalence<Array<A>> =>
  Eq.make((as, bs) =>
    pipe(
      as,
      A.matchLeft({
        onEmpty: () => A.isEmptyArray(bs),
        onNonEmpty: (a, as_) =>
          pipe(
            A.findFirstIndex(bs, Eq.equals(E)(a)),
            O.map(i => A.remove(i)(bs)),
            O.match({
              onNone: () => false,
              onSome: Eq.equals(getUnorderedEquivalence(E))(as_),
            }),
          ),
      }),
    ),
  )

export const getPermutations: <A>(as: Array<A>) => Array<Array<A>> = A.match({
  onEmpty: constant([[]]),
  onNonEmpty: as =>
    pipe(
      as,
      A.map((a, i) =>
        pipe(A.remove(i)(as), getPermutations, A.map(A.append(a))),
      ),
      A.flatten,
    ),
})
