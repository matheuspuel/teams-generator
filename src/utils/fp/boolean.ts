import { Endomorphism } from 'fp-ts/Endomorphism'
export * from 'fp-ts/boolean'

export const toggle: Endomorphism<boolean> = value => !value
