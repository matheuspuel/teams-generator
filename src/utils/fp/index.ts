export * as Chunk from '@effect/data/Chunk'
export * as Context from '@effect/data/Context'
export * as Data from '@effect/data/Data'
export * as Duration from '@effect/data/Duration'
export * as Predicate from '@effect/data/Predicate'
export * as Tuple from '@effect/data/Tuple'
export * as Clock from '@effect/io/Clock'
export * as F from '@effect/io/Effect'
export * as Layer from '@effect/io/Layer'
export * as LogLevel from '@effect/io/LogLevel'
export * as Logger from '@effect/io/Logger'
export * as Ref from '@effect/io/Ref'
export * as Runtime from '@effect/io/Runtime'
export * as SynchronizedRef from '@effect/io/SynchronizedRef'
export * as Stream from '@effect/stream/Stream'
export * as SubscriptionRef from '@effect/stream/SubscriptionRef'
export * as Console from 'fp-ts/Console'
export * as A from './Array'
export * as Boolean from './Boolean'
export * as E from './Either'
export * as Eq from './Eq'
export * as Match from './Match'
export * as Monoid from './Monoid'
export * as Number from './Number'
export * as Optic from './Optic'
export * as O from './Option'
export * as Ord from './Order'
export * as Reader from './Reader'
export * as Record from './Record'
export * as S from './Schema'
export * as Semigroup from './Semigroup'
export * as String from './String'

export {
  LazyArg,
  absurd,
  apply,
  constFalse,
  constNull,
  constTrue,
  constUndefined,
  constVoid,
  constant,
  flip,
  flow,
  identity,
  pipe,
} from '@effect/data/Function'
export { and, not, or } from '@effect/data/Predicate'
export { Effect } from '@effect/io/Effect'
export { Endomorphism } from 'fp-ts/Endomorphism'
export { NonEmptyArray } from 'fp-ts/NonEmptyArray'
export { State } from 'fp-ts/State'
export { Either } from './Either'
export { get, modify, replace } from './Optic'
export { Option } from './Option'
export { Order } from './Order'
