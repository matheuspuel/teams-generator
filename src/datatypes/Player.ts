import {
  Array,
  Boolean,
  Number,
  Option,
  Order,
  Schema,
  String,
  flow,
  identity,
  pipe,
} from 'effect'
import { NonEmptyReadonlyArray } from 'effect/Array'
import { sumAll } from 'effect/Number'
import { t } from 'src/i18n'
import { Id } from 'src/utils/Entity'
import { normalize } from 'src/utils/String'
import { Timestamp } from 'src/utils/datatypes'
import { average } from 'src/utils/fp/Number'
import { Modality } from './Modality'
import * as Position from './Position'
import * as Rating from './Rating'

export class Player extends Schema.Class<Player>('Player')({
  id: Id,
  name: Schema.String,
  rating: Rating.Rating,
  positionAbbreviation: Position.Abbreviation,
  active: Schema.Boolean,
  createdAt: Timestamp.Timestamp,
}) {}

export const isActive = (p: Player) => p.active

const getPositionAndIndex =
  (args: { modality: Modality }) =>
  (
    player: Player,
  ): Option.Option<{ position: Position.Position; index: number }> =>
    pipe(
      identity<
        NonEmptyReadonlyArray<Position.StaticPosition | Position.CustomPosition>
      >(args.modality.positions),
      Array.findFirst((po, i) =>
        po.abbreviation === player.positionAbbreviation
          ? Option.some({ position: po, index: i })
          : Option.none(),
      ),
    )

export const PositionOrd = (args: {
  modality: Modality
}): Order.Order<Player> =>
  Order.mapInput(Number.Order, p =>
    pipe(
      getPositionAndIndex(args)(p),
      Option.match({ onNone: () => -1, onSome: _ => _.index }),
    ),
  )

export const NameOrd: Order.Order<Player> = pipe(
  String.Order,
  Order.mapInput(flow(p => p.name, normalize)),
)

export const RatingOrd: Order.Order<Player> = pipe(
  Number.Order,
  Order.mapInput(p => p.rating),
)

export const ActiveOrd: Order.Order<Player> = pipe(
  Boolean.Order,
  Order.mapInput(p => p.active),
)

export const CreatedAtOrder: Order.Order<Player> = pipe(
  Timestamp.Order,
  Order.mapInput(p => p.createdAt),
)

export const position =
  (args: { modality: Modality }) =>
  (player: Player): Option.Option<Position.Position> =>
    Option.map(getPositionAndIndex(args)(player), _ => _.position)

export const toString =
  (args: { modality: Modality }): ((player: Player) => string) =>
  p =>
    `${Rating.toString(p.rating)} - ${p.name} (${pipe(
      position(args)(p),
      Option.map(Position.toAbbreviationString),
      Option.getOrElse(() => '-'),
    )})`

export const listToString =
  (args: { modality: Modality }) =>
  (players: Array<Player>): string =>
    pipe(
      players,
      Array.sortBy(PositionOrd(args), Order.reverse(RatingOrd), NameOrd),
      Array.map(toString(args)),
      Array.join('\n'),
    )

export const teamListToString =
  (args: { modality: Modality }) =>
  (teams: Array<Array<Player>>): string =>
    pipe(
      teams,
      Array.map(listToString(args)),
      Array.map((v, i) => `${t('Team')} ${i + 1}\n\n${v}`),
      Array.join('\n\n'),
    )

export const toStringSensitive =
  (args: { modality: Modality }) =>
  (player: Player): string =>
    `${player.name} (${pipe(
      position(args)(player),
      Option.map(Position.toAbbreviationString),
      Option.getOrElse(() => '-'),
    )})`

export const listToStringSensitive =
  (args: { modality: Modality }) =>
  (players: Array<Player>): string =>
    pipe(
      players,
      Array.sortBy(PositionOrd(args), NameOrd),
      Array.map(toStringSensitive(args)),
      Array.join('\n'),
    )

export const teamListToStringSensitive =
  (args: { modality: Modality }) =>
  (teams: Array<Array<Player>>): string =>
    pipe(
      teams,
      Array.map(listToStringSensitive(args)),
      Array.map((v, i) => `${t('Team')} ${i + 1}\n\n${v}`),
      Array.join('\n\n'),
    )

export const getRatingTotal: (players: Array<Player>) => number = players =>
  sumAll(Array.map(players, p => p.rating))

export const getRatingAverage: (
  players: Array<Player>,
) => Option.Option<number> = flow(
  Array.map(p => p.rating),
  average,
)
