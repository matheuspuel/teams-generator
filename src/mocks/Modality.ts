import { Modality } from 'src/datatypes'
import { Abbreviation } from 'src/datatypes/Position'
import { Id } from 'src/utils/Entity'
import { NonEmptyString } from 'src/utils/datatypes/NonEmptyString'

export const soccerMock: Modality = {
  id: Id('1'),
  name: NonEmptyString('Futebol'),
  positions: [
    {
      id: Id('G'),
      abbreviation: Abbreviation('g'),
      name: NonEmptyString('Goleiro'),
    },
    {
      id: Id('Z'),
      abbreviation: Abbreviation('z'),
      name: NonEmptyString('Zagueiro'),
    },
    {
      id: Id('LE'),
      abbreviation: Abbreviation('le'),
      name: NonEmptyString('Lateral Esquerdo'),
    },
    {
      id: Id('LD'),
      abbreviation: Abbreviation('ld'),
      name: NonEmptyString('Lateral Direito'),
    },
    {
      id: Id('V'),
      abbreviation: Abbreviation('v'),
      name: NonEmptyString('Volante'),
    },
    {
      id: Id('M'),
      abbreviation: Abbreviation('m'),
      name: NonEmptyString('Meia'),
    },
    {
      id: Id('A'),
      abbreviation: Abbreviation('a'),
      name: NonEmptyString('Atacante'),
    },
  ],
}
