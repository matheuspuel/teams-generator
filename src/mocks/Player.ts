import { Player } from 'src/datatypes'
import { Id } from 'src/utils/Entity'
import { Timestamp } from 'src/utils/datatypes'

export const playersMock: Array<Player> = [
  {
    id: '1' as Id,
    name: 'Matheus',
    rating: 10,
    position: 'M',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '2' as Id,
    name: 'Paulo A.',
    rating: 4,
    position: 'A',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '3' as Id,
    name: 'Carlos',
    rating: 5,
    position: 'Z',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '4' as Id,
    name: 'Paulo G.',
    rating: 8,
    position: 'G',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '5' as Id,
    name: 'Peter',
    rating: 6,
    position: 'Z',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '6' as Id,
    name: 'Zeca',
    rating: 3,
    position: 'LE',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '7' as Id,
    name: 'Moisés',
    rating: 3,
    position: 'LD',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '8' as Id,
    name: 'Odilon',
    rating: 9,
    position: 'M',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '9' as Id,
    name: 'Paulo S.',
    rating: 10,
    position: 'A',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '10' as Id,
    name: 'João',
    rating: 7,
    position: 'M',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '11' as Id,
    name: 'Anderson',
    rating: 3,
    position: 'Z',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '12' as Id,
    name: 'Gilmar',
    rating: 6,
    position: 'LD',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '13' as Id,
    name: 'Vagner',
    rating: 8,
    position: 'M',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '14' as Id,
    name: 'Douglas',
    rating: 2,
    position: 'LD',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '15' as Id,
    name: 'Marlon',
    rating: 4,
    position: 'A',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '16' as Id,
    name: 'Marcos',
    rating: 8,
    position: 'M',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '17' as Id,
    name: 'Jackson',
    rating: 4,
    position: 'A',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '18' as Id,
    name: 'Wagner',
    rating: 7,
    position: 'A',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '19' as Id,
    name: 'Moa',
    rating: 5,
    position: 'A',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '20' as Id,
    name: 'Neto',
    rating: 5,
    position: 'A',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '21' as Id,
    name: 'Lucas',
    rating: 7,
    position: 'Z',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '22' as Id,
    name: 'Jefferson',
    rating: 3,
    position: 'LD',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
  {
    id: '23' as Id,
    name: 'Clóvis',
    rating: 7,
    position: 'Z',
    active: true,
    createdAt: Timestamp.Schema(0),
  },
]
