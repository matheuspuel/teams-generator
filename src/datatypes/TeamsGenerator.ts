import {
  Array,
  Effect,
  Iterable,
  Match,
  Number,
  Option,
  Order,
  Tuple,
  flow,
  identity,
  pipe,
} from 'effect'
import { NonEmptyReadonlyArray } from 'effect/Array'
import { sumAll } from 'effect/Number'
import * as Player from 'src/datatypes/Player'
import * as Position from 'src/datatypes/Position'
import { randomizeArray } from 'src/utils/Random'
import { Modality } from './Modality'

type Player = Player.Player
type Position = Position.Position

const getFitOrdByDevianceFns = (
  fns: Array<(teams: Array<Array<Player>>) => number>,
): Order.Order<Array<Array<Player>>> =>
  pipe(
    fns,
    Array.map(f => Order.mapInput(f)(Number.Order)),
    Order.combineAll<Array<Array<Player>>>,
  )

const getResultPositionDeviance =
  (args: { modality: Modality }) =>
  (teams: Array<Array<Player>>): number =>
    pipe(teams, Array.flatten, allPlayers =>
      pipe(
        identity<
          NonEmptyReadonlyArray<
            Position.StaticPosition | Position.CustomPosition
          >
        >(args.modality.positions),
        Array.map(pos =>
          pipe(
            allPlayers,
            positionCount(pos),
            n => n / teams.length,
            positionAverage =>
              pipe(
                teams,
                Array.map(flow(positionCount(pos), deviance(positionAverage))),
                sumAll,
              ),
          ),
        ),
        sumAll,
      ),
    )

const fixFloatFactor = 1000000000000

const fixFloat = (v: number) => Math.round(v * fixFloatFactor) / fixFloatFactor

export const getResultRatingDeviance = (teams: Array<Array<Player>>): number =>
  pipe(
    teams,
    Array.flatten,
    Player.getRatingAverage,
    Option.getOrElse(() => 0),
    overallAverage =>
      pipe(
        teams,
        Array.map(
          flow(
            Player.getRatingAverage,
            Option.getOrElse(() => 0),
            deviance(overallAverage),
          ),
        ),
        sumAll,
        fixFloat,
      ),
  )

const balanceTeams = (args: {
  teams: Array<Array<Player>>
  fitOrder: Order.Order<Array<Array<Player>>>
}): Effect.Effect<Array<Array<Player>>> =>
  Effect.sync(() => {
    const isMoreBalanced = Order.lessThan(args.fitOrder)
    let teams = args.teams
    // eslint-disable-next-line no-constant-condition
    repeatTryingAllPossibleSimpleChanges: while (true) {
      for (const [team1, team1Index] of Iterable.map(teams, Tuple.make)) {
        for (const [team2, team2Index] of Iterable.map(teams, Tuple.make)) {
          if (team2Index <= team1Index) continue
          for (const player1Index of Iterable.map(team1, (_, i) => i)) {
            for (const player2Index of Iterable.map(team2, (_, i) => i)) {
              const changedTeams = changePlayers(
                { teamIndex: team1Index, playerIndex: player1Index },
                { teamIndex: team2Index, playerIndex: player2Index },
              )(teams)
              if (isMoreBalanced(changedTeams, teams)) {
                teams = changedTeams
                continue repeatTryingAllPossibleSimpleChanges
              }
            }
          }
        }
      }
      break
    }
    return teams
  })

const deviance = (b: number) => (a: number) => Math.pow(Math.abs(a - b), 2)

const positionCount =
  (position: Position) =>
  (players: Array<Player>): number =>
    pipe(
      players,
      Array.filter(p => p.positionAbbreviation === position.abbreviation),
      Array.length,
    )

const changePlayers =
  (
    first: { teamIndex: number; playerIndex: number },
    second: { teamIndex: number; playerIndex: number },
  ) =>
  (teams: Array<Array<Player>>): Array<Array<Player>> =>
    Option.gen(function* () {
      const firstTeam = yield* Array.get(teams, first.teamIndex)
      const secondTeam = yield* Array.get(teams, second.teamIndex)
      const firstPlayer = yield* Array.get(firstTeam, first.playerIndex)
      const secondPlayer = yield* Array.get(secondTeam, second.playerIndex)
      const nextFirstTeam = pipe(
        firstTeam,
        Array.remove(first.playerIndex),
        Array.append(secondPlayer),
      )
      const nextSecondTeam = pipe(
        secondTeam,
        Array.remove(second.playerIndex),
        Array.append(firstPlayer),
      )
      return yield* pipe(
        teams,
        Array.replaceOption<Array<Player>>(first.teamIndex, nextFirstTeam),
        Option.flatMap(
          Array.replaceOption<Array<Player>>(second.teamIndex, nextSecondTeam),
        ),
      )
    }).pipe(Option.getOrElse(() => teams))

export const divideTeamsWithEqualNumberOfPlayers = (args: {
  players: Array<Player>
  numOfTeams: number
}): Array<Array<Player>> =>
  args.numOfTeams <= 0
    ? []
    : pipe(
        args.players,
        Array.splitAt(Math.floor(args.players.length / args.numOfTeams)),
        ([as, bs]) =>
          pipe(
            divideTeamsWithEqualNumberOfPlayers({
              players: bs,
              numOfTeams: args.numOfTeams - 1,
            }),
            Array.append(as),
          ),
      )

export const divideTeamsWithFixedNumberOfPlayers = (args: {
  players: Array<Player>
  numOfPlayers: number
}): Array<Array<Player>> =>
  args.players.length === 0
    ? []
    : pipe(args.players, Array.splitAt(args.numOfPlayers), ([as, bs]) =>
        pipe(
          divideTeamsWithFixedNumberOfPlayers({
            players: bs,
            numOfPlayers: args.numOfPlayers,
          }),
          Array.prepend(as),
        ),
      )

const divideTeams = (args: { players: Array<Player>; criteria: Criteria }) =>
  Match.valueTags(args.criteria.distribution, {
    numOfTeams: ({ numOfTeams }) =>
      divideTeamsWithEqualNumberOfPlayers({
        players: args.players,
        numOfTeams,
      }),
    fixedNumberOfPlayers: ({ fixedNumberOfPlayers }) =>
      divideTeamsWithFixedNumberOfPlayers({
        players: args.players,
        numOfPlayers: fixedNumberOfPlayers,
      }),
  })

export type Criteria = {
  position: boolean
  rating: boolean
  distribution:
    | { _tag: 'numOfTeams'; numOfTeams: number }
    | { _tag: 'fixedNumberOfPlayers'; fixedNumberOfPlayers: number }
}

export const getFitOrdFromCriteria = (args: {
  modality: Modality
  criteria: Criteria
}): Order.Order<Array<Array<Player>>> =>
  pipe(
    [
      args.criteria.position
        ? Option.some(getResultPositionDeviance(args))
        : Option.none(),
      args.criteria.rating
        ? Option.some(getResultRatingDeviance)
        : Option.none(),
    ],
    Array.getSomes,
    getFitOrdByDevianceFns,
  )

export const balanceTeamsByCriteria = (args: {
  teams: Array<Array<Player>>
  modality: Modality
  criteria: Criteria
}) => balanceTeams({ teams: args.teams, fitOrder: getFitOrdFromCriteria(args) })

export const distributeTeams = (args: {
  players: Array<Player>
  modality: Modality
  criteria: Criteria
}) => balanceTeamsByCriteria({ ...args, teams: divideTeams(args) })

export const generateRandomBalancedTeams = (args: {
  players: Array<Player>
  modality: Modality
  criteria: Criteria
}): Effect.Effect<Array<Array<Player>>> =>
  Effect.gen(function* () {
    return yield* distributeTeams({
      ...args,
      players: yield* randomizeArray(args.players),
    })
  })
