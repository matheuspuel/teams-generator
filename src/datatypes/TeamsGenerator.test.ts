/* eslint-disable functional/no-expression-statements */
import * as Arb from '@effect/schema/Arbitrary'
import * as fc from 'fast-check'
import { $, A, Monoid, Num } from 'fp'
import { playersMock } from 'src/mocks/Player'
import { Player } from '.'
import { balanceTeamsByCriteria } from './TeamsGenerator'

describe('Balance teams', () => {
  it('should return the same number of players', () => {
    fc.assert(
      fc.property(
        fc.integer({ min: 1, max: 20 }),
        fc.boolean(),
        fc.boolean(),
        fc.array(Arb.to(Player.Schema)(fc)),
        (n, position, rating, players) =>
          Monoid.concatAll(Num.MonoidSum)(
            balanceTeamsByCriteria({ position, rating })(n)(players).map(
              A.size,
            ),
          ) === players.length,
      ),
    )
  })

  it('should return the correct number of teams', () => {
    fc.assert(
      fc.property(
        fc.integer({ min: 1, max: 20 }),
        fc.boolean(),
        fc.boolean(),
        fc.array(Arb.to(Player.Schema)(fc)),
        (n, position, rating, players) =>
          balanceTeamsByCriteria({ position, rating })(n)(players).length === n,
      ),
    )
  })

  it('should not have teams with player count difference higher than one', () => {
    fc.assert(
      fc.property(
        fc.integer({ min: 1, max: 20 }),
        fc.boolean(),
        fc.boolean(),
        fc.array(Arb.to(Player.Schema)(fc)),
        (n, position, rating, players) =>
          $(balanceTeamsByCriteria({ position, rating })(n)(players), teams =>
            teams.every(a =>
              teams.every(b => Math.abs(a.length - b.length) <= 1),
            ),
          ),
      ),
    )
  })

  it('should return a balanced team', () => {
    const result = balanceTeamsByCriteria({ position: true, rating: true })(3)(
      playersMock,
    )
    expect(result).toMatchSnapshot()
  })
})
