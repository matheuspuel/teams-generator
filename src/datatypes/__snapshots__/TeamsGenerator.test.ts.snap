// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Balance teams should return a balanced team 1`] = `
[
  [
    {
      "active": true,
      "id": "15",
      "name": "Marlon",
      "position": "A",
      "rating": 4,
    },
    {
      "active": true,
      "id": "2",
      "name": "Paulo A.",
      "position": "A",
      "rating": 4,
    },
    {
      "active": true,
      "id": "13",
      "name": "Vagner",
      "position": "M",
      "rating": 8,
    },
    {
      "active": true,
      "id": "4",
      "name": "Paulo G.",
      "position": "G",
      "rating": 8,
    },
    {
      "active": true,
      "id": "12",
      "name": "Gilmar",
      "position": "LD",
      "rating": 6,
    },
    {
      "active": true,
      "id": "3",
      "name": "Carlos",
      "position": "Z",
      "rating": 5,
    },
    {
      "active": true,
      "id": "23",
      "name": "Clóvis",
      "position": "Z",
      "rating": 7,
    },
    {
      "active": true,
      "id": "17",
      "name": "Jackson",
      "position": "A",
      "rating": 4,
    },
  ],
  [
    {
      "active": true,
      "id": "9",
      "name": "Paulo S.",
      "position": "A",
      "rating": 10,
    },
    {
      "active": true,
      "id": "11",
      "name": "Anderson",
      "position": "Z",
      "rating": 3,
    },
    {
      "active": true,
      "id": "14",
      "name": "Douglas",
      "position": "LD",
      "rating": 2,
    },
    {
      "active": true,
      "id": "16",
      "name": "Marcos",
      "position": "M",
      "rating": 8,
    },
    {
      "active": true,
      "id": "18",
      "name": "Wagner",
      "position": "A",
      "rating": 7,
    },
    {
      "active": true,
      "id": "10",
      "name": "João",
      "position": "M",
      "rating": 7,
    },
    {
      "active": true,
      "id": "21",
      "name": "Lucas",
      "position": "Z",
      "rating": 7,
    },
    {
      "active": true,
      "id": "22",
      "name": "Jeferson",
      "position": "LD",
      "rating": 3,
    },
  ],
  [
    {
      "active": true,
      "id": "1",
      "name": "Matheus",
      "position": "M",
      "rating": 10,
    },
    {
      "active": true,
      "id": "5",
      "name": "Peter",
      "position": "Z",
      "rating": 6,
    },
    {
      "active": true,
      "id": "6",
      "name": "Zeca",
      "position": "LE",
      "rating": 3,
    },
    {
      "active": true,
      "id": "7",
      "name": "Moisés",
      "position": "LD",
      "rating": 3,
    },
    {
      "active": true,
      "id": "19",
      "name": "Moa",
      "position": "A",
      "rating": 5,
    },
    {
      "active": true,
      "id": "20",
      "name": "Neto",
      "position": "A",
      "rating": 5,
    },
    {
      "active": true,
      "id": "8",
      "name": "Odilon",
      "position": "M",
      "rating": 9,
    },
  ],
]
`;
